x/gamm/keeper/invariants.go

```diff
@@ -78,3 +86,65 @@ func PoolTotalWeightInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant
 			fmt.Sprintf("\tgamm all pool calculated and stored total weight match\n")), false
 	}
 }
+
+func genericPow(base, exp sdk.Dec) sdk.Dec {
+	if !base.GTE(sdk.NewDec(2)) {
+		return pow(base, exp)
+	}
+	return powApprox(sdk.OneDec().Quo(base), exp.Neg(), powPrecision)
+}
+
+// constantChange returns the multiplicative factor difference in the pool constant, between two different pools.
+// For a Balancer pool, the pool constant is prod_{t in tokens} t.bal^{t.weight}
+func constantChange(p1, p2 types.PoolI) sdk.Dec {
+	product := sdk.OneDec()
+	totalWeight := p1.GetTotalWeight()
+	assets1 := p1.GetAllPoolAssets()
+	for _, asset1 := range assets1 {
+		asset2, err := p2.GetPoolAsset(asset1.Token.Denom)
+		if err != nil {
+			panic(err)
+		}
+		w := asset1.Weight.ToDec().Quo(totalWeight.ToDec())
+		ratio := asset1.Token.Amount.ToDec().Quo(asset2.Token.Amount.ToDec())
+		power := genericPow(ratio, w)
+		product = product.Mul(power)
+	}
+
+	return product
+}
+
+var (
+	errorMargin, _ = sdk.NewDecFromStr("0.0001") // 0.01%
+)
+
+func PoolProductConstantInvariant(keeper Keeper) sdk.Invariant {
```

Comment:
Nice, this is a good way of doing it!

This should fail with liquidity removal though, correct?

Reply:
Also can you comment the function

Reply:
Yes, it should fail with liquidity removal.

Reply:
Lets add a coment explicitly stating that, and a TODO for ourselves to revisit how to do this with liquidity removal? (Perhaps we have a simulator-only variable for trackign amount of liquidity removed since last invokation)